__tests__/Auth0Client.test.ts

```diff
@@ -173,28 +173,389 @@ describe('Auth0Client', () => {
+        expect.objectContaining({
+          refresh_token: TEST_REFRESH_TOKEN
+        })
+      );
+    });
+    it('should save refresh_token in local storage cache', async () => {
+      const auth0 = setup({
+        useRefreshTokens: true,
+        cacheLocation: 'localstorage'
+      });
+
+      jest.spyOn(auth0['cache'], 'save');
+
+      await loginWithPopup(auth0);
+
+      expect(auth0['cache']['save']).toHaveBeenCalledWith(
+        expect.objectContaining({
+          refresh_token: TEST_REFRESH_TOKEN
+        })
+      );
+    });
+
+    it('saves `auth0.is.authenticated` key in storage', async () => {
+      const auth0 = setup();
+
+      jest.spyOn(auth0['cookieStorage'], 'save');
+
+      await loginWithPopup(auth0);
+
+      expect(auth0['cookieStorage']['save']).toHaveBeenCalledWith(
+        'auth0.is.authenticated',
+        true,
+        {
+          daysUntilExpire: 1
+        }
+      );
+    });
+    it('saves `auth0.is.authenticated` key in storage for an extended period', async () => {
+      const auth0 = setup({
+        sessionCheckExpiryDays: 2
+      });
+
+      jest.spyOn(auth0['cookieStorage'], 'save');
+
+      await loginWithPopup(auth0);
+
+      expect(auth0['cookieStorage']['save']).toHaveBeenCalledWith(
+        'auth0.is.authenticated',
+        true,
+        {
+          daysUntilExpire: 2
+        }
+      );
+    });
+
+    it('should throw an error on token failure', async () => {
+      const auth0 = setup();
+
+      await expect(
+        loginWithPopup(auth0, {}, {}, false, {})
+      ).rejects.toThrowError(
+        'HTTP error. Unable to fetch https://auth0_domain/oauth/token'
+      );
+    });
+  });
+
   describe('loginWithRedirect', () => {
     it('should log the user in and get the token', async () => {
       const auth0 = setup();
       await loginWithRedirect(auth0);
       const url = new URL(mockWindow.location.assign.mock.calls[0][0]);
-      assertUrlEquals(url, 'auth0_domain', '/authorize', {
-        client_id: 'auth0_client_id',
-        redirect_uri: 'my_callback_url',
-        scope: 'openid profile email',
+      assertUrlEquals(url, TEST_DOMAIN, '/authorize', {
+        client_id: TEST_CLIENT_ID,
+        redirect_uri: TEST_REDIRECT_URI,
+        scope: TEST_SCOPES,
         response_type: 'code',
         response_mode: 'query',
-        state: 'MTIz',
-        nonce: 'MTIz',
-        code_challenge: '',
+        state: TEST_STATE,
+        nonce: TEST_NONCE,
+        code_challenge: TEST_CODE_CHALLENGE,
         code_challenge_method: 'S256'
       });
       assertPost('https://auth0_domain/oauth/token', {
-        redirect_uri: 'my_callback_url',
-        client_id: 'auth0_client_id',
-        code_verifier: '123',
+        redirect_uri: TEST_REDIRECT_URI,
+        client_id: TEST_CLIENT_ID,
+        code_verifier: TEST_CODE_VERIFIER,
         grant_type: 'authorization_code',
-        code: 'my_code'
+        code: TEST_CODE
```

Comment:
I also updated existing `loginWithRedirect` tests to use the const values I introduced to avoid randon strings.